src/coreclr/src/vm/classlayoutinfo.cpp

```diff
@@ -0,0 +1,1105 @@
+
+    BOOL fParentHasLayout = pParentMT && pParentMT->HasLayout();
+    UINT32 cbAdjustedParentLayoutNativeSize = 0;
+    EEClassNativeLayoutInfo const* pParentLayoutInfo = NULL;
+    if (fParentHasLayout)
+    {
+        pParentLayoutInfo = pParentMT->GetNativeLayoutInfo();
+        // Treat base class as an initial member.
+        cbAdjustedParentLayoutNativeSize = pParentLayoutInfo->GetSize();
+        // If the parent was originally a zero-sized explicit type but
+        // got bumped up to a size of 1 for compatibility reasons, then
+        // we need to remove the padding, but ONLY for inheritance situations.
+        if (pParentMT->GetLayoutInfo()->IsZeroSized()) {
+            CONSISTENCY_CHECK(cbAdjustedParentLayoutNativeSize == 1);
+            cbAdjustedParentLayoutNativeSize = 0;
+        }
+    }
+
+
+    CorNativeLinkType charSet = pMT->GetCharSet();
+
+    ParseNativeTypeFlags nativeTypeFlags = ParseNativeTypeFlags::None;
+#ifdef FEATURE_COMINTEROP
+    if (pMT->IsProjectedFromWinRT())
+        nativeTypeFlags = ParseNativeTypeFlags::IsWinRT;
+    else // WinRT types have nlType == nltAnsi but should be treated as Unicode
+#endif // FEATURE_COMINTEROP
+        if (charSet == nltAnsi)
+            nativeTypeFlags = ParseNativeTypeFlags::IsAnsi;
+
+    ApproxFieldDescIterator fieldDescs(pMT, ApproxFieldDescIterator::INSTANCE_FIELDS);
+
+    ULONG cInstanceFields = fieldDescs.Count();
+
+    NewArrayHolder<LayoutRawFieldInfo> pInfoArray = new LayoutRawFieldInfo[cInstanceFields + 1];
+
+    SigTypeContext context(pMT);
+
+    ParseFieldNativeTypes(
+        pInternalImport,
+        pMT->GetCl(),
+        fieldDescs,
+        pModule,
+        nativeTypeFlags,
+        &context,
+        pInfoArray
+        DEBUGARG(szNamespace)
+        DEBUGARG(szName)
+    );
+
+    uint32_t numTotalInstanceFields = cInstanceFields + (pParentLayoutInfo != nullptr ? pParentLayoutInfo->GetNumFields() : 0);
+    LoaderAllocator* pAllocator = pMT->GetLoaderAllocator();
+    AllocMemHolder<EEClassNativeLayoutInfo> pNativeLayoutInfo(
+        pAllocator->GetLowFrequencyHeap()->AllocMem(
+            S_SIZE_T(sizeof(EEClassNativeLayoutInfo)) + S_SIZE_T(sizeof(NativeFieldDescriptor)) * S_SIZE_T(numTotalInstanceFields)));
+
+    pNativeLayoutInfo->m_numFields = numTotalInstanceFields;
+
+    // Now compute the native size of each field
+    for (LayoutRawFieldInfo* pfwalk = pInfoArray; pfwalk->m_MD != mdFieldDefNil; pfwalk++)
+    {
+        pfwalk->m_placement.m_size = pfwalk->m_nfd.NativeSize();
+        pfwalk->m_placement.m_alignment = pfwalk->m_nfd.AlignmentRequirement();
+    }
+
+    S_UINT32 cbSortArraySize = S_UINT32(cInstanceFields) * S_UINT32(sizeof(LayoutRawFieldInfo*));
+    if (cbSortArraySize.IsOverflow())
+    {
+        ThrowHR(COR_E_TYPELOAD);
+    }
+
+    BOOL fExplicitOffsets = pMT->GetClass()->HasExplicitFieldOffsetLayout();
+
+    LayoutRawFieldInfo** pSortArray = (LayoutRawFieldInfo**)_alloca(cbSortArraySize.Value());
```

Comment:
I know you have not introduced this `_alloca`, but it is potentially very large (ie can be 100kB for auto-generated code). `CQuickArray<LayoutRawFieldInfo*>` would be more appropriate.